File: examples/brick breaker/brick_breaker.cpp
#include "goopylib/goopylib.h"


class Brick : public gp::Rectangle {
public:
    Brick(int i, int j)
            : gp::Rectangle({i * s_Width - 450 + s_Width / 2,
                             j * s_Height + 85}, s_Width, s_Height) {

        setColor(gp::Color(255 - j * 20, 255 - j * 20, 255 - j * 20));
    }

    void hit() {
        m_Hits += 1;
        setTransparency(1 - 0.25f * m_Hits);

        if (m_Hits == 4) {
            destroy();
        }
    }

    int m_Hits = 0;

    static const int s_Width = 100;
    static const int s_Height = 55;
};

class Ball : public gp::Circle {
public:
    Ball() : gp::Circle({-250, 20}, 7) {
        setColor(gp::Color(255, 150, 150));
    }

    void update(float dt) {
        if (m_Position.x < -450) {
            setX(-450);
            m_xVel *= -1;
        }
        else if (m_Position.x > 450) {
            setX(450);
            m_xVel *= -1;
        }

        if (m_Position.y > 300) {
            setY(300);
            collide();
        }

        if (abs(m_yVel) < 5) {
            m_yVel -= 0.07;
            m_xVel += 0.07;
        }

        move(70 * m_xVel * dt, 70 * m_yVel * dt);
    }

    void collide() {
        m_yVel *= -1;
    }

    float m_xVel = 0;
    float m_yVel = 0;
};

class Controller : public gp::Rectangle {
public:
    Controller() : gp::Rectangle({0, -220}, 120, 15) {
        setColor(gp::Color(255, 255, 255));
    }

    void update() {
        auto pos = m_Window->toWorld(m_Window->getMousePosition());

        float tmp = m_Position.x;
        setX(max(-m_MaxX, min(m_MaxX, pos.x)));
        m_xVel = m_Position.x - tmp;
    }

    float m_xVel = 0;
    float m_MaxX = 450;
};


int main(int argc, char* argv[]) {
    gp::init();

    auto window = gp::Window(900, 600, "Brick Breaker using goopylib!");
    window.setBackground(gp::Color(25, 25, 25));
    window.setResizable(true);

    window.setMinSize(600, 400);
    window.setAspectRatio(900, 600);

    auto &camera = window.getCamera();

    std::string argv_str(argv[0]);
    std::string base = argv_str.substr(0, argv_str.find_last_of("/"));

    std::string path = base + "/vignette.png";

    auto vignette = gp::Image(path.c_str(), {0, 0}, 900, 600);
    vignette.draw(&window);
    vignette.setTransparency(0.3, 0.3, 0.8, 0.8);

    while (window.isOpen()) {
        camera.setRotation(0);
        camera.setZoom(1);

        auto controller = Controller();
        controller.draw(&window);

        auto ball = Ball();
        ball.draw(&window);

        float transparency = 0;

        std::vector<Brick> bricks;

        for (int i = 0; i < 900 / Brick::s_Width; i++) {
            for (int j = 0; j < 4; j++) {
                bricks.push_back(Brick(i, j));
                bricks.back().draw(&window);
            }
        }

        while (window.isOpen() and !window.checkLeftClick()) {
            if (transparency < 1) {
                transparency += 0.02;
                ball.setTransparency(min(1.0f, transparency));
            }

            controller.update();
            gp::update();
        }

        ball.setTransparency(1);

        float dt = 0;
        float lasthit = 0;

        while (window.isOpen()) {
            float start = gp::getTime();

            if (ball.getY() < -300 or bricks.size() == 0) {
                break;
            }

            if (controller.contains(ball.getPosition())) {
                ball.collide();
                ball.m_xVel = max(-12.0f, min(12.0f, ball.m_xVel * 0.85f + controller.m_xVel * 0.2f));
                ball.move(0, 12);
            }

            for (auto &brick: bricks) {
                if (brick.isDrawn() and brick.contains(ball.getPosition())) {
                    ball.collide();
                    brick.hit();
                    lasthit = gp::getTime();
                    break;
                }
            }

            if (gp::getTime() - lasthit < 0.2) {
                camera.setX(rand() % 8 - 4);
            }
            else {
                camera.setX(0);
            }

            controller.update();
            ball.update(dt);
            gp::update();

            dt = gp::getTime() - start;
        }

        ball.destroy();
        auto start = gp::getTime();
        auto ease = gp::easeBounceOut(4, 0.8);
        float easet = 0.7;

        controller.m_MaxX = 350;

        while (window.isOpen() and gp::getTime() - start < easet) {
            float val = ease(gp::getTime() - start);
            camera.setRotation(4 * val);
            camera.setZoom(1 - 0.02 * val);

            controller.update();
            gp::update();
        }

        while (window.isOpen() and !window.checkLeftClick()) {
            controller.update();
            gp::update();
        }

        while (window.isOpen() and window.checkLeftClick()) {
            controller.update();
            gp::update();
        }

        while (window.isOpen() and camera.getRotation() > 0) {
            camera.rotate(-0.2);
            camera.setZoom(1 - 0.005 * camera.getRotation());

            controller.update();
            gp::update();
        }

        controller.destroy();

        for (auto &brick: bricks) {
            brick.destroy();
        }
    }

    gp::terminate();
    return 0;
}

File: examples/brick breaker/brick_breaker.py
import goopylib.imports as gp

import random
import time
import os

PATH = os.path.dirname(os.path.realpath(__file__))


class Brick(gp.Image):
    width = 100
    height = 55

    colors = ["blue", "purple", "yellow", "green"]

    def __init__(self, i, j):
        super().__init__(f"{PATH}/assets/element_{self.colors[j]}_rectangle.png",
                         (i * self.width - 450 + self.width / 2, j * self.height + 85),
                         self.width, self.height)
        self.hits = 0

    def hit(self):
        self.hits += 1
        self.set_transparency(1 - 0.25 * self.hits)


class Ball(gp.Circle):
    def __init__(self, p):
        super().__init__(p, 10)
        self.set_color((255, 255, 255))
        
        self.velx = 0 if p[0] < 0 else -5
        self.vely = -7

        self.lastcontrollerx = 0

    def update(self, dt, controller, blocks):
        if self.x < -450 or self.x > 450:
            self.velx *= -1
        if self.y < -300 or self.y > 300:
            self.vely *= -1

        velx = controller.x - self.lastcontrollerx
        self.lastcontrollerx = controller.x

        if controller.contains(*self.position):
            self.velx = max(-12, min(12, self.velx * 0.85 + velx * 0.2))
            self.vely *= -1
            self.y = controller.y + 12

        for block in blocks:
            if block.contains(*self.position):
                self.vely *= -1
                block.hit()

                if block.hits == 4:
                    block.destroy()
                    blocks.remove(block)

        self.move(70 * self.velx * dt, 70 * self.vely * dt)

        if self.y < -300:
            return -1
        return 0


def main():
    window = gp.Window(900, 600, "Brick Breaker using goopylib!")
    window.background = (10, 5, 50)
    gp.set_buffer_swap_interval(0)
    window.resizable = True

    window.set_min_size(600, 400)
    window.set_aspect_ratio(900, 600)

    gp.Image(f"{PATH}/assets/background.png", (0, 0)).draw(window).set_transparency(0.4)

    controller = gp.Rectangle((0, -220), 120, 20).draw(window)
    controller.set_color((255, 255, 255))
    ball = Ball((random.randint(-450, 450), 20)).draw(window)

    blocks = []
    for i in range(900 // Brick.width):
        for j in range(4):
            blocks.append(Brick(i, j).draw(window))

    dt = 0.01

    while window.is_open():
        start = time.time()

        mousex, _ = window.to_world(*window.get_mouse_position())

        controller.x = max(-450, min(450, mousex))
        end = ball.update(dt, controller, blocks)

        if end or len(blocks) == 0:
            for i in range(100):
                controller.set_transparency(1 - i * 0.01)
                ball.set_transparency(1 - i * 0.01)
                for block in blocks:
                    block.set_transparency(1 - i * 0.01)

                time.sleep(0.005)
                gp.update()

            break

        gp.update()

        time.sleep(max(0, 0.01 - (time.time() - start)))


if __name__ == "__main__":
    main()

gp.terminate()
